CRM/Civicase/Service/CaseTypeCategoryEventHandler.php

```diff
@@ -0,0 +1,236 @@
+<?php
+
+/**
+ * Class CRM_Civicase_Service_CaseTypeCategoryEventHandler.
+ */
+class CRM_Civicase_Service_CaseTypeCategoryEventHandler {
+
+  /**
+   * Menu handler.
+   *
+   * @var \CRM_Civicase_Service_CaseCategoryMenu
+   */
+  protected $menu;
+
+  /**
+   * Custom data handler.
+   *
+   * @var \CRM_Civicase_Service_CaseCategoryCustomDataType
+   */
+  protected $customData;
+
+  /**
+   * Custom field handler.
+   *
+   * @var \CRM_Civicase_Service_CaseCategoryCustomFieldExtends
+   */
+  protected $customFieldExtends;
+
+  /**
+   * Should we check referer before running action?
+   *
+   * @var bool
+   */
+  protected $needCheckReferer = FALSE;
+
+  /**
+   * Should we check option group before running action?
+   *
+   * Depends on $optionGroupId.
+   *
+   * @var bool
+   */
+  protected $needCheckOptionGroup = TRUE;
+
+  /**
+   * Case type category option group id. Used to check option group.
+   *
+   * @var int
+   */
+  protected $optionGroupId;
+
+  /**
+   * Object name.
+   *
+   * Is used to check object if it's type is unpredictable or unknown.
+   *
+   * @var string
+   */
+  protected $objectName = 'OptionValue';
+
+  /**
+   * CRM_Civicase_Service_CaseTypeCategoryEventHandler constructor.
+   *
+   * @param object $menu
+   *   Menu handler.
+   * @param object $customData
+   *   Custom data handler.
+   * @param object $customFieldExtends
+   *   Custom field handler.
+   */
+  public function __construct($menu, $customData, $customFieldExtends) {
```

Comment:
Can we add the actual class rather than just `object` for the parameters and then type hint the parameters these function expects to reflect these classes.